feat(travis): Add explicit Travis integration

jervi opened the pull request:
This will add a new Travis trigger and a new Travis stage.
Travis has been supported in Spinnaker for a while, but has confusingly shared configuration with Jenkins in Deck. This commit will split the two implementations.
Corresponds with https://github.com/spinnaker/echo/pull/141 and https://github.com/spinnaker/orca/pull/1213.

BREAKING CHANGE: Existing pipelines will continue to work, but `travis` will have to be added to `triggerTypes` in `settings.js` before you can set up new travis triggers/stages or edit any existing travis triggers/stages.

tomaslin commented on app/scripts/modules/core/ci/jenkins/igor.service.js:
```diff
@@ -8,10 +8,11 @@ module.exports = angular.module('spinnaker.core.ci.jenkins.igor.service', [
 
     function listMasters() {
-      return API.one('v2').one('builds').get();
+      return API.one('v2').one('builds').get()
```
Why not just keep one igorService and have a filter `listMasters('jenkins')` and `listMasters('travis')`

tomaslin replied:
reason for this is that the two services can now easily go out of sync easily

jervi replied:
Agree 👍 I've updated the PR.

danielpeach commented on halconfig/settings.js:
```diff
@@ -105,7 +105,7 @@ window.spinnakerSettings = {
   authTtl: 600000,
   gitSources: ['stash', 'github', 'bitbucket'],
-  triggerTypes: ['git', 'pipeline', 'docker', 'cron', 'jenkins'],
+  triggerTypes: ['git', 'pipeline', 'docker', 'cron', 'jenkins', 'travis'],
```
Could you please make a corresponding update to the settings.js files under the https://github.com/spinnaker/spinnaker repo?

jervi replied:
Yes, I will fix it :)
- [ ] Update spinnaker repo

jervi replied:
@danielpeach This line does not exist in that file; should I add it?
https://github.com/spinnaker/spinnaker/blob/master/config/settings.js#L122

danielpeach commented on app/scripts/modules/core/pipeline/config/triggers/travis/travisTriggerOptions.directive.js:
```diff
@@ -0,0 +1,64 @@
+  .module('spinnaker.core.pipeline.config.triggers.travis.options.directive', [
+    require('core/ci/igor.service.js')
+  ])
+  .directive('travisTriggerOptions', function () {
```
I am not sure if the corresponding `jenkinsTriggerOptions` is a directive, but it would be preferable if this were a component (it doesn't look like it needs to be a directive). 

The syntax:

```
.component('travisTriggerOptions`, { 
  templateUrl: ...,
  bindings: { command: '='},
  controller: ...
})
```

`$ctrl` is the default value for `controllerAs` for a component.

jervi commented:
Thanks @danielpeach :) Most of the implementation is copy/paste from the existing Jenkins implementation, that's why it's still Javascript. But I can take a shot at converting it to TypeScript.

icfantv commented on app/scripts/modules/core/pipeline/config/stages/travis/travisStage.module.js:
```diff
@@ -0,0 +1,12 @@
+'use strict';
```
This should be in Typescript

jervi commented on app/scripts/modules/core/pipeline/config/stages/travis/travisExecutionDetails.controller.js:
```diff
@@ -0,0 +1,49 @@
+      }
+
+      if (failingTestCount) {
+        failureMessage = failingTestCount + ' test(s) failed.';
```
This is how it is done in the Jenkins implementation. I guess you'll have to visit Travis/Jenkins to get the details.

jervi commented:
Sorry about the slow response, and thank you for all the great feedback. I've converted all *.js classes to TypeScript. I also converted the IgorService and added some new domain classes. I hope this looks better. Feedback is appreciated :)
cc/ @tomaslin @danielpeach @icfantv

icfantv commented:
@jervi no worries.  i'll re-review tonight.

icfantv commented:
@jervi i apologize.  i will get to this this weekend.

icfantv commented on app/scripts/modules/core/pipeline/config/stages/travis/travisStage.ts:
```diff
@@ -0,0 +1,174 @@
+}
+
+export class TravisStage {
+  static get $inject() {
```
Move to be next to the ctor.

icfantv commented on app/scripts/modules/core/pipeline/config/triggers/travis/selectorTemplate.html:
```diff
@@ -0,0 +1 @@
+<travis-trigger-options command="vm.command"></travis-trigger-options>
```
This should be in-lined as it's one line of HTML rather than creating a separate file.

jervi commented on app/scripts/modules/core/pipeline/config/triggers/travis/travisTrigger.module.ts:
```diff
@@ -0,0 +1,126 @@
+    formatLabel: (trigger: IBuildTrigger) => {
+      return $q.when(`(Travis) ${trigger.master}: ${trigger.job}`);
+    },
+    selectorTemplate: '<travis-trigger-options command="$ctrl.command"></travis-trigger-options>',
```
@icfantv I've inlined the html. Is the command parameter correct? I couldn't figure out how to test it. It was `vm.command` in the html file.

icfantv commented:
@jervi oh, and please squash your commits.

jervi commented:
I've reverted to `vm`, squashed and rebased. Thank you for the feedback, @icfantv 👍

icfantv commented:
@jervi awesome.  let's merge this thing.
